File: lib/rag/generation.ts
import { GoogleGenAI } from "@google/genai";
import { config } from "@/lib/config";
import { GenerationError } from "@/lib/errors";
import { buildUserMessage, SYSTEM_PROMPT } from "@/lib/rag/prompt";
import type { RetrievedChunk } from "@/lib/rag/retrieval";

/**
 * Gemini answer generation. This is the ONLY module that imports the Google
 * Gen AI SDK. Answers are streamed so the UI renders tokens as they arrive.
 *
 * Grounding rules are supplied via the system instruction; the retrieved
 * chunks and question are supplied as the user content. Temperature is 0 for
 * faithful, deterministic grounding, and model thinking is disabled to keep
 * short factual answers fast.
 *
 * @module lib/rag/generation
 */

let client: GoogleGenAI | null = null;

function getClient(): GoogleGenAI {
  if (!config.geminiApiKey) {
    throw new GenerationError(
      "GEMINI_API_KEY is not configured. Set it in .env.local (or .env).",
    );
  }
  if (!client) {
    client = new GoogleGenAI({ apiKey: config.geminiApiKey });
  }
  return client;
}

/**
 * Stream a grounded answer for a question and its retrieved context.
 *
 * @param question - The user's question.
 * @param chunks - Retrieved context chunks.
 * @yields Answer text fragments as they arrive.
 * @throws {GenerationError} When the API key is missing or the request fails.
 */
export async function* streamAnswer(
  question: string,
  chunks: RetrievedChunk[],
): AsyncGenerator<string> {
  const ai = getClient();
  try {
    const stream = await ai.models.generateContentStream({
      model: config.geminiModel,
      contents: buildUserMessage(question, chunks),
      config: {
        systemInstruction: SYSTEM_PROMPT,
        maxOutputTokens: config.geminiMaxTokens,
        temperature: 0,
        thinkingConfig: { thinkingBudget: 0 },
      },
    });

    for await (const chunk of stream) {
      const text = chunk.text;
      if (text) yield text;
    }
  } catch (error) {
    if (error instanceof GenerationError) throw error;
    const detail = error instanceof Error ? error.message : "unknown error";
    throw new GenerationError(`Answer generation failed: ${detail}`);
  }
}

File: backend/app/services/generation.py
from collections.abc import Iterable, Iterator
from functools import lru_cache

from google import genai
from google.genai import types

from app.config import settings
from app.grounding import SYSTEM_PROMPT, build_user_message
from app.services.retrieval import RetrievedChunk

"""Gemini answer generation. The only module that streams from the generation
model. Grounding rules go in the system instruction; retrieved chunks and the
question go in the contents. Temperature 0 and thinking disabled keep grounded
answers faithful and fast."""


class GenerationError(Exception):
    """Raised when generation fails (e.g. missing API key or upstream error)."""


@lru_cache(maxsize=1)
def _client() -> genai.Client:
    key = settings.resolved_gemini_key
    if not key:
        raise GenerationError("GEMINI_API_KEY is not configured. Set it in .env.")
    return genai.Client(api_key=key)


def stream_answer(
    question: str, chunks: Iterable[RetrievedChunk]
) -> Iterator[str]:
    """Stream a grounded answer for a question and its retrieved context."""
    client = _client()
    try:
        stream = client.models.generate_content_stream(
            model=settings.gemini_model,
            contents=build_user_message(question, chunks),
            config=types.GenerateContentConfig(
                system_instruction=SYSTEM_PROMPT,
                max_output_tokens=settings.gemini_max_tokens,
                temperature=0,
                thinking_config=types.ThinkingConfig(thinking_budget=0),
            ),
        )
        for chunk in stream:
            if chunk.text:
                yield chunk.text
    except GenerationError:
        raise
    except Exception as error:  # noqa: BLE001 - surfaced as a domain error
        raise GenerationError(f"Answer generation failed: {error}") from error
